odl/discr/discr_ops.py

```diff
@@ -54,74 +61,81 @@ def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=2,
+        provided, the returned object is a reference to it.
 
     Examples
     --------
     >>> f = np.array([ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
+
     >>> finite_diff(f)
     array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
-    >>> finite_diff(f, axis=0, dx=1.0, edge_order=2, zero_padding=False)
+
+    Without arguments the above defaults to:
+    >>> finite_diff(f, axis=0, dx=1.0, method='forward', padding=None,
+    ... edge_order=None)
     array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
+
     >>> finite_diff(f, dx=0.5)
     array([ 2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.])
-    >>> finite_diff(f, zero_padding=True)
-    array([ 0.5,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. , -4. ])
-    >>> finite_diff(f, zero_padding=False, edge_order=1)
-    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
-    >>> out = finite_diff(f)
-    >>> out is finite_diff(f)
-    False
+    >>> finite_diff(f, padding=0)
+    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1., -9.])
+
+    Central differences and different edge orders
+    >>> finite_diff(1/2*f**2, method='central')
+    array([-0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.])
+    >>> finite_diff(1/2*f**2, method='central', edge_order=1)
+    array([ 0.5,  1. ,  2. ,  3. ,  4. ,  5. ,  6. ,  7. ,  8. ,  8.5])
+
+    In-place evaluation:
+    >>> out = f.copy()
     >>> out is finite_diff(f, out)
     True
     """
     # TODO: implement alternative boundary conditions
 
-    if zero_padding is True and edge_order == 1:
-        raise ValueError("zero padding uses second-order accurate "
-                         "differences at boundaries. First-order accurate "
-                         "edges can only be used without zero padding.")
-    f_data = np.asarray(f)
-    ndim = f_data.ndim
+    f_arr = np.asarray(f)
+    ndim = f_arr.ndim
 
-    if not 0 <= axis < ndim:
-        raise IndexError("axis parameter ({0}) exceeds the number of "
-                         "dimensions ({1}).".format(axis, ndim))
-
-    if f_data.shape[axis] < 2:
+    if f_arr.shape[axis] < 2:
         raise ValueError("shape ({0}) of array too small to calculate a "
                          "numerical gradient, at least two elements are "
-                         "required.".format(f_data.shape))
+                         "required.".format(f_arr.shape))
 
     if out is None:
-        out = np.empty_like(f_data)
+        out = np.empty_like(f_arr)
     else:
         if out.shape != f.shape:
             raise ValueError(
                 "shape of `out` array ({0}) does not match the shape of "
                 "input array `f` ({1}).".format(out.shape, f.shape))
 
-    if edge_order not in [1, 2]:
-        raise ValueError("edge order ({0}) not valid".format(edge_order))
+    if not 0 <= axis < ndim:
+        raise IndexError("axis parameter ({0}) exceeds the number of "
+                         "dimensions ({1}).".format(axis, ndim))
```

Comment:
Allow negative indices, simply by adding `ndim` in that case. Error message: like `'axis {} outside the valid range 0 ... {}'.format(axis, ndim - 1)`